src/dune_rules/pkg_rules.ml

```diff
@@ -621,6 +623,109 @@ module Substitute = struct
 end
 
 module Run_with_path = struct
+  module Output : sig
+    type t
+
+    val io : t -> Process.Io.output Process.Io.t
+
+    val v
+      :  ?accepted_exit_codes:int Predicate.t
+      -> ?pkg_name:Dune_lang.Package_name.t
+      -> string
+      -> t
+
+    val consume_and_print_error : t -> Display.t -> code:int -> loc:Loc.t -> unit
+  end = struct
+    type state =
+      | Open
+      | Closed
+
+    type t =
+      { pkg_name : Dune_pkg.Package_name.t option
+      ; fn : Dpath.t
+      ; io : Process.Io.output Process.Io.t
+      ; accepted_exit_codes : int Predicate.t
+      ; mutable state : state
+      }
+
+    let prefix = "dune-pkg"
+    let exit_code_zero = Predicate.create (fun x -> x = 0)
+    let has_output error = error <> ""
+
+    let v ?(accepted_exit_codes = exit_code_zero) ?pkg_name suffix =
+      let fn = Temp.create File ~prefix ~suffix in
+      let io = Process.Io.(file fn Out) in
+      let state = Open in
+      { pkg_name; fn; io; state; accepted_exit_codes }
+    ;;
+
+    let io t = t.io
+
+    let close t =
+      t.state <- Closed;
+      Temp.destroy File t.fn
+    ;;
+
+    let read t =
+      match t.state with
+      | Open ->
+        let content = Stdune.Io.read_file t.fn in
+        content
+      | Closed ->
+        raise
+          (User_error.E
+             (User_message.make
+                [ Pp.text "Run_with_path: impossible to read data from a closed state" ]))
+    ;;
+
+    let error_msg ~loc report_type pkg_name error =
+      let open Pp.O in
+      let report_msg, report_style =
+        match report_type with
+        | `Warning -> Pp.text "Warning", User_message.Style.Warning
+        | `Error -> Pp.text "Error", User_message.Style.Error
+      in
+      let pp_package =
+        match pkg_name with
+        | None -> Pp.nop
+        | Some pkg_name ->
+          let pkg_name = Dune_pkg.Package_name.to_string pkg_name in
+          Pp.(
+            tag report_style (text "<><><>")
```

Comment:
no need to mimic what opam is doing here. this clutters output and is less accessible than just printing the error message.